File: src/main/java/com/machfour/macros/cli/Total.java
package com.machfour.macros.cli;

import com.machfour.macros.linux.Config;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDatabase;

import java.io.PrintStream;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.cli.CliMain.OUT;
import static com.machfour.macros.cli.CliMain.PROGNAME;

class Total extends CommandImpl {
    private static final String NAME = "total";
    @Override
    public String name() {
        return NAME;
    }
    @Override
    public void printHelp(PrintStream out) {
        OUT.printf("Usage: %s %s (<meal name>|--all) [<day>] [-v|--verbose] [--per100]\n", PROGNAME, NAME);
    }

    private static MealSpec makeMealSpec(List<String> args, boolean isAllMeals) {
        /* logic:
         * if isAllMeals is true:
         *     day is the first thing that doesn't start with -- or -
         * if isAllMeals is false:
         *     meal name is the first thing that doesn't start with -- or -
         *     day is the second such thing
         */
        List<String> nonOptionArgs = new ArrayList<>(args.size() - 1);
        for (String arg : args) {
            // add everything after the first arg (which is the mode name) which doesn't start with a -
            if (!(arg.equals(args.get(0)) || arg.startsWith("-"))) {
                nonOptionArgs.add(arg);
            }
        }
        String mealName = null;
        String dayString = null;
        if (nonOptionArgs.size() >= 1) {
            if (isAllMeals) {
                // just look for day
                dayString = nonOptionArgs.get(0);
            } else {
                // look for day and meal
                mealName = nonOptionArgs.get(0);
                if (nonOptionArgs.size() >= 2) {
                    dayString = nonOptionArgs.get(1);
                }
            }
        }
        return MealSpec.makeMealSpec(mealName, dayString);
    }
    @Override
    public void doAction(List<String> args) {
        if (args.contains("--help")) {
            printHelp(OUT);
            return;
        }
        boolean verbose = args.contains("--verbose") || args.contains("-v");
        boolean per100 = args.contains("--per100");
        boolean allMeals = args.contains("--all");

        MacrosDatabase db = LinuxDatabase.getInstance(Config.DB_LOCATION);
        MealSpec spec = makeMealSpec(args, allMeals);
        process(spec, db, allMeals, verbose, per100);

    }

    static void process(MealSpec mealSpec, MacrosDatabase db, boolean allMeals, boolean verbose, boolean per100) {
        if (!allMeals) {
            // total for specific meal
            mealSpec.process(db, false);
            if (mealSpec.error() != null) {
                OUT.println(mealSpec.error());
                return;
            }
            OUT.println();
            MealPrinter.printMeal(mealSpec.processedObject(), verbose, OUT);

        } else {
            try {
                Map<String, Meal> mealsForDay = db.getMealsForDay(mealSpec.day());
                if (mealsForDay.isEmpty()) {
                    OUT.println("No meals recorded on " + CliUtils.prettyDay(mealSpec.day()));
                } else {
                    MealPrinter.printMeals(mealsForDay.values(), OUT, verbose, per100, true);
                }
            } catch (SQLException e) {
                OUT.println("Error retrieving meals: " + e.getMessage());
                OUT.println();
            }
        }
    }

}

File: macros-cli/src/main/kotlin/com/machfour/macros/cli/modes/Total.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CliConfig
import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.makeMealSpec
import com.machfour.macros.cli.utils.printMeal
import com.machfour.macros.cli.utils.printMeals
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.parsing.MealSpec
import com.machfour.macros.queries.getMealsForDay
import com.machfour.macros.sql.SqlDatabase
import com.machfour.macros.sql.SqlException

/*
 * Prints out totals for all DB recorded meals in a day
 */
class Total(config: CliConfig) : CommandImpl(config) {
    override val name = "total"
    override val usage = "Usage: ${config.programName} $name (<meal name>|--all) [<day>] [-v|--verbose] [--per100]"

    companion object {
        private fun makeMealSpec(args: List<String>, isAllMeals: Boolean): MealSpec {
            /* logic:
             * if isAllMeals is true:
             *     day is the first thing that doesn't start with -- or -
             * if isAllMeals is false:
             *     meal name is the first thing that doesn't start with -- or -
             *     day is the second such thing
             */
            val nonOptionArgs = args.filter { it != args.firstOrNull() && !it.startsWith("-") }
            val mealName = if (nonOptionArgs.isNotEmpty() && !isAllMeals) nonOptionArgs[0] else null
            val dayString = when {
                nonOptionArgs.isEmpty() -> null
                isAllMeals -> nonOptionArgs[0] // just look for day
                nonOptionArgs.size >= 2 -> nonOptionArgs[1] // look for meal and day
                else -> null
            }

            return makeMealSpec(mealName, dayString)
        }
    }

    override fun doAction(args: List<String>): Int {
        if (args.contains("--help")) {
            printHelp()
            return 0
        }
        val verbose = args.contains("--verbose") || args.contains("-v")
        val per100 = args.contains("--per100")
        val allMeals = args.contains("--all")

        val ds = config.database
        val spec = makeMealSpec(args, allMeals)
        return process(spec, ds, allMeals, verbose, per100)

    }

    private fun process(mealSpec: MealSpec, ds: SqlDatabase, allMeals: Boolean, verbose: Boolean, per100: Boolean): Int {
        if (!allMeals) {
            // total for specific meal
            mealSpec.process(ds, false)
            if (mealSpec.error != null) {
                printlnErr(mealSpec.error)
                return 1
            }
            println()
            printMeal(mealSpec.processedObject!!, verbose)

        } else {
            try {
                val day = mealSpec.day!!
                val mealsForDay = getMealsForDay(ds, day)
                if (mealsForDay.isEmpty()) {
                    println("No meals recorded on " + day.prettyPrint())
                } else {
                    printMeals(mealsForDay.values, verbose, per100, true)
                }
            } catch (e: SqlException) {
                println()
                printlnErr("Error retrieving meals: " + e.message)
                return 1
            }

        }
        return 0
    }


}
